### PR TITLE
add top to job requests api

### LINKED ISSUE #4324: Azure Pipelines trigger failing with "Client.Timeout exceeded while awaiting headers"
### Report

I configured the Azure Pipelines trigger (with `scaledJob`) as explained [here](https://keda.sh/docs/2.4/scalers/azure-pipelines/) for few small pools in my organization and it worked well.
When I configured it to another pool that running many pipelines the scaling didn't work. 
I checked Keda logs and saw: 
```
{"error": "Get \"https://my-azure-devops/_apis/distributedtask/pools/52/jobrequests\": context deadline exceeded (Client.Timeout exceeded while awaiting headers)"}
```
When I tried to access the above api URL I succeeded to get results, but - it took 4 seconds because the result count is 760.

image image

So I guess this is the issue and because it takes 4 seconds it failed with `Client.Timeout exceeded`.

### Expected Behavior

Scaling up/down of the `scaledJob` according to the waiting pipelines.

### Actual Behavior

Nothing happened, new Jobs wasn't created. 

### Steps to Reproduce the Problem

1. Configure `scaledJob` with Azure DevOps pool that running **many** pipelines.
2. Keda not auto scaling.

### Logs from KEDA operator

```
{"error": "Get \"https://my-azure-devops/_apis/distributedtask/pools/52/jobrequests\": context deadline exceeded (Client.Timeout exceeded while awaiting headers)"}
```

### KEDA Version

2.9.3

### Kubernetes Version

< 1.23

### Platform

Other

### Scaler Details

Azure Pipelines

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by shayki5**
_Provide a description of what has been changed_

### Checklist

- [x] When introducing a new scaler, I agree with the [scaling governance policy](https://github.com/kedacore/governance/blob/main/SCALERS.md)
- [x] I have verified that my change is according to the [deprecations & breaking changes policy](https://github.com/kedacore/governance/blob/main/DEPRECATIONS.md)
- [x] Tests have been added
- [x] Changelog has been updated and is aligned with our [changelog requirements](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#Changelog)
- [x] A PR is opened to update the documentation on ([repo](https://github.com/kedacore/keda-docs)) *(if applicable)*
- [x] Commits are signed with Developer Certificate of Origin (DCO - [learn more](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#developer-certificate-of-origin-signing-your-work))

Fixes #4324

### CODE REVIEW

**Comment by JorTurFer** [comment]
Hi @shayki5 
I ❤️  this improvement! 
Should we allow users to set their own values for the top (keeping 250 or whatever as default)? I mean, end-users can increase the timeout using env vars and maybe there are some user cases in really large orgs where 250 isn't enough, maybe giving the option to set the top value (in addition to the already existing option for increasing the timeout) would fulfil that requirement.
WDYT?

**Comment by shayki5** [comment]
Hi @JorTurFer, thank you for your quick response! 
Yea, it's a good idea, it could be a variable, I just need to check how to do it...
I thought 250 it's enough because need more than 250 waiting **jobs** in one pool for the scale not to work properly, but as you said, maybe in large orgs it's should be the case.

**Review by zroubalik — COMMENTED** [review]
I agree with @JorTurFer, we should add a property for this :) Thanks for the contribution!

File: pkg/scalers/azure_pipelines_scaler.go [hunk 1]
```diff
@@ -217,6 +218,12 @@ func parseAzurePipelinesMetadata(ctx context.Context, config *ScalerConfig, http
 	} else {
 		meta.demands = ""
 	}
+	
+	if val, ok := config.TriggerMetadata["jobsToFetch"]; ok && val != "" {
+		meta.jobsToFetch = config.TriggerMetadata["jobsToFetch"]
+	} else {
+		meta.jobsToFetch = 250
+	}
```
Comment (JorTurFer) [inline pkg/scalers/azure_pipelines_scaler.go:226]:
```suggestion
	meta.jobsToFetch = 250
	if val, ok := config.TriggerMetadata["jobsToFetch"]; ok && val != "" {
		meta.jobsToFetch = config.TriggerMetadata["jobsToFetch"]
	} 
```

**Review by JorTurFer — COMMENTED** [review]
Looking good! Could you add this new field to the parsing tests?
As we are adding a new parameter, a PR to docs is also required explaining the new parameter. This is the file you need to edit https://github.com/kedacore/keda-docs/blob/main/content/docs/2.10/scalers/azure-pipelines.md

**Comment by shayki5** [comment]
@JorTurFer Sure, thank you so much for the parsing test, I added it.
I opened a PR for the docs: https://github.com/kedacore/keda-docs/pull/1082

**Comment by JorTurFer** [comment]
> Sure, thank you so much for the parsing test, I added it.

I didn't add the parsing test, I just suggested a change in the code to follow the sytle xD You still need to add a case here https://github.com/kedacore/keda/blob/main/pkg/scalers/azure_pipelines_scaler_test.go#L26

**Comment by shayki5** [comment]
> > Sure, thank you so much for the parsing test, I added it.
> 
> I didn't add the parsing test, I just suggested a change in the code to follow the sytle xD You still need to add a case here https://github.com/kedacore/keda/blob/main/pkg/scalers/azure_pipelines_scaler_test.go#L26

Oh sorry, my bad, this is what happens that today is the first day for you to write go ;) I just followed the `demands` syntax...
Which test should I add? I see the `testAzurePipelinesMetadata` checks if there are empty values but in this case the `jobsToFetch` it's optional, it can be empty.

**Comment by JorTurFer** [comment]
> Which test should I add?

I think that we should cover 3 cases, empty (already covered xD), correctly provided and incorrectly provided. I mean, in case of empty we don't have anything to test, but we should test that if a correct value is set we don't raise any error and in case of wrong value set, we raise an error.

**Comment by shayki5** [comment]
> > Which test should I add?
> 
> I think that we should cover 3 cases, empty (already covered xD), correctly provided and incorrectly provided. I mean, in case of empty we don't have anything to test, but we should test that if a correct value is set we don't raise any error and in case of wrong value set, we raise an error. This can be done just adding the new 2 cases to the array

I understand, I tried to look in the existing tests but couldn't figure out how can I implement those tests :(

**Comment by JorTurFer** [comment]
You can check how this case (activationTargetPipelinesQueueLength malformed) is added: https://github.com/kedacore/keda/blob/main/pkg/scalers/azure_pipelines_scaler_test.go#L41-L42

Basically, you need to provide the scaler values ([you can take this line as base](https://github.com/kedacore/keda/blob/main/pkg/scalers/azure_pipelines_scaler_test.go#L30)) including the new parameter, in one case wrongly (for example, setting "a" instead of a number) and other case correctly. If you check the line options, there is a boolean to specify if the case should raise an error or not

**Comment by shayki5** [comment]
> You can check how this case (activationTargetPipelinesQueueLength malformed) is added: https://github.com/kedacore/keda/blob/main/pkg/scalers/azure_pipelines_scaler_test.go#L41-L42
> 
> Basically, you need to provide the scaler values ([you can take this line as base](https://github.com/kedacore/keda/blob/main/pkg/scalers/azure_pipelines_scaler_test.go#L30)) including the new parameter, in one case wrongly (for example, setting "a" instead of a number) and other case correctly. If you check the line options, there is a boolean to specify if the case should raise an error or not. The wrong case should have true in that field because it should raise an error, the correct case should have false because not error should be raised

Thank you so much for your explanation! I added something, can you check..?

File: CHANGELOG.md [hunk 2]
```diff
@@ -81,6 +81,7 @@ Here is an overview of all new **experimental** features:
 - **Azure Queue Scaler**: Fix azure queue length ([#4002](https://github.com/kedacore/keda/issues/4002))
 - **Azure Service Bus Scaler:** Use correct auth flows with pod identity ([#4026](https://github.com/kedacore/keda/issues/4026)|[#4123](https://github.com/kedacore/keda/issues/4123))
 - **Azure Service Bus Scaler:**: Improve way clients are created to reduce amount of ARM requests ([#4262](https://github.com/kedacore/keda/issues/4262))
+- **Azure Pipelines Scaler**: Fix scaling in big pools. ([#4324](https://github.com/kedacore/keda/issues/4324))
```
Comment (JorTurFer) [inline CHANGELOG.md:84]: Our changelog style places `General` section on top, an the rest alphabetically sorted

```suggestion
- **Azure Pipelines Scaler**: Add new parameter to limit the jobs returned ([#4324](https://github.com/kedacore/keda/issues/4324))
- **Azure Queue Scaler**: Fix azure queue length ([#4002](https://github.com/kedacore/keda/issues/4002))
- **Azure Service Bus Scaler:** Use correct auth flows with pod identity ([#4026](https://github.com/kedacore/keda/issues/4026)|[#4123](https://github.com/kedacore/keda/issues/4123))
- **Azure Service Bus Scaler:**: Improve way clients are created to reduce amount of ARM requests ([#4262](https://github.com/kedacore/keda/issues/4262))
```

**Review by JorTurFer — APPROVED** [review]
LGTM! Thanks for the improvement.
Please update the changelog (The comment I wrote)

File: pkg/scalers/azure_pipelines_scaler_test.go [hunk 3]
```diff
@@ -40,6 +40,9 @@ var testAzurePipelinesMetadata = []parseAzurePipelinesMetadataTestData{
 	{"missing poolID", map[string]string{"organizationURLFromEnv": "AZP_URL", "personalAccessTokenFromEnv": "AZP_TOKEN", "poolID": "", "targetPipelinesQueueLength": "1"}, true, testAzurePipelinesResolvedEnv, map[string]string{}},
 	// activationTargetPipelinesQueueLength malformed
 	{"all properly formed", map[string]string{"organizationURLFromEnv": "AZP_URL", "personalAccessTokenFromEnv": "AZP_TOKEN", "poolID": "1", "targetPipelinesQueueLength": "1", "activationTargetPipelinesQueueLength": "A"}, true, testAzurePipelinesResolvedEnv, map[string]string{}},
+	// jobsToFetch malformed
+	{"jobsToFetch malformed", map[string]string{"organizationURLFromEnv": "AZP_URL", "personalAccessTokenFromEnv": "AZP_TOKEN", "poolID": "1", "targetPipelinesQueueLength": "1", "jobsToFetch": "test"}, false, testAzurePipelinesResolvedEnv, map[string]string{}},
```
Comment (JorTurFer) [inline pkg/scalers/azure_pipelines_scaler_test.go:44]: This case has to raise an error

```suggestion
	{"jobsToFetch malformed", map[string]string{"organizationURLFromEnv": "AZP_URL", "personalAccessTokenFromEnv": "AZP_TOKEN", "poolID": "1", "targetPipelinesQueueLength": "1", "jobsToFetch": "test"}, true, testAzurePipelinesResolvedEnv, map[string]string{}},
```

File: pkg/scalers/azure_pipelines_scaler_test.go [hunk 4]
```diff
@@ -40,6 +40,9 @@ var testAzurePipelinesMetadata = []parseAzurePipelinesMetadataTestData{
 	{"missing poolID", map[string]string{"organizationURLFromEnv": "AZP_URL", "personalAccessTokenFromEnv": "AZP_TOKEN", "poolID": "", "targetPipelinesQueueLength": "1"}, true, testAzurePipelinesResolvedEnv, map[string]string{}},
 	// activationTargetPipelinesQueueLength malformed
 	{"all properly formed", map[string]string{"organizationURLFromEnv": "AZP_URL", "personalAccessTokenFromEnv": "AZP_TOKEN", "poolID": "1", "targetPipelinesQueueLength": "1", "activationTargetPipelinesQueueLength": "A"}, true, testAzurePipelinesResolvedEnv, map[string]string{}},
+	// jobsToFetch malformed
+	{"jobsToFetch malformed", map[string]string{"organizationURLFromEnv": "AZP_URL", "personalAccessTokenFromEnv": "AZP_TOKEN", "poolID": "1", "targetPipelinesQueueLength": "1", "jobsToFetch": "test"}, true, testAzurePipelinesResolvedEnv, map[string]string{}},
+
 }
```
Comment (JorTurFer) [inline pkg/scalers/azure_pipelines_scaler_test.go:46]: And the last one, static checks are failing
```suggestion
	// jobsToFetch malformed
	{"jobsToFetch malformed", map[string]string{"organizationURLFromEnv": "AZP_URL", "personalAccessTokenFromEnv": "AZP_TOKEN", "poolID": "1", "targetPipelinesQueueLength": "1", "jobsToFetch": "test"}, true, testAzurePipelinesResolvedEnv, map[string]string{}},
}
```

**Comment by shayki5** [comment]
Fixed :) thank you so much!!

**Review by zroubalik — APPROVED** [review]

**Comment by JorTurFer** [comment]
Thanks for the contribution!
This feature will be released as part of v2.10

**Comment by shayki5** [comment]
Your prompt response and coordination are greatly appreciated!
Do you know when v2.10 will be released?

**Comment by JorTurFer** [comment]
> Do you know when v2.10 will be released?

Most probably this week

**Pull request merged**